packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,762 @@
+                        if let DependencySource::StateVar {
+                            component_idx,
+                            state_var_idx,
+                        } = &dep.source
+                        {
+                            let new_component_state = ComponentStateDescription {
+                                component_idx: *component_idx,
+                                state_var_idx: *state_var_idx,
+                            };
+
+                            let new_current_freshness = dep.value.get_freshness();
+
+                            match new_current_freshness {
+                                // No need to recurse if the state var of the dependency is already fresh
+                                // so don't do anything to just continue the inner loop
+                                Freshness::Fresh => (),
+                                Freshness::Stale => {
+                                    // Recurse by putting two items on the stack
+                                    // and then continuing at start of stack loop
+                                    freshen_stack.push(StateVarCalculationState::Stale(
+                                        StaleCalculationState {
+                                            component_state,
+                                            instruction_idx,
+                                            val_idx,
+                                        },
+                                    ));
+                                    freshen_stack.push(StateVarCalculationState::Stale(
+                                        StaleCalculationState {
+                                            component_state: new_component_state,
+                                            instruction_idx: 0,
+                                            val_idx: 0,
+                                        },
+                                    ));
+
+                                    // break out of the inner loop and continue to outer loop,
+                                    // which will result in first popping off the last instruction
+                                    // that we added to the stack
+                                    continue 'stack_loop;
+                                }
+                                Freshness::Unresolved => {
+                                    panic!("How did a stale state variable depend on an unresolved state variable?")
+                                }
+                            };
+                        }
+                    }
+
+                    initial_val_idx = 0;
+                }
+
+                // If we have gotten to this part,
+                // then we have freshened all dependencies
+                // for the state variable (component_idx, state_var_idx).
+                // We calculate its value if a dependency has changed,
+                // else restore its previous value.
```

Comment:
I must have missed it. Where is the actual calculation to "freshen" a var done?

Oh, I see, it's only the _dependencies_ which we can assume have been freshend at this point. The real freshening is below.

Reply:
Right, the freshening is done by `calculate_state_var_from_dependencies`. Maybe I should rename that to `calculate_state_var_from_dependencies_and_mark_fresh`?

Reply:
I renamed it in case that helps our future selves.